core/hybrisadaptor.cpp

```diff
@@ -1228,59 +1248,152 @@ void *HybrisManager::eventReaderThread(void *aptr)
             do { } while (nanosleep(&ts, &ts) == -1 && errno == EINTR);
             continue;
         }
-        /* Process received events */
-        int blockSuspend = 0;
-        bool errorInInput = false;
-        manager->processEvents(buffer, numberOfEvents, blockSuspend, errorInInput);
-
-        /* Suspend proof sensor reporting that could occur in display off */
-        if (blockSuspend) {
-            ObtainTemporaryWakeLock();
-        }
+        // Queue received events for processing in main thread
+        manager->queueEvents(buffer, numberOfEvents);
 #endif
-        /* Rate limit after receiving erraneous events */
-        if (errorInInput) {
-            struct timespec ts = { 0, 50 * 1000 * 1000 }; // 50 ms
-            do { } while (nanosleep(&ts, &ts) == -1 && errno == EINTR);
-        }
     }
     return 0;
 }
 
-void HybrisManager::processEvents(const sensors_event_t *buffer, int numberOfEvents, int &blockSuspend, bool &errorInInput)
+void HybrisManager::initEventPipe()
 {
-    for (int i = 0; i < numberOfEvents; i++) {
-        const sensors_event_t& data = buffer[i];
+    sensordLogD("initialize event pipe");
+    int pfd[2] = {-1, -1};
+    if (::pipe2(pfd, O_CLOEXEC) == -1) {
+        sensordLogW("failed to create event pipe: %s", strerror(errno));
+    } else {
+        m_eventPipeReadFd = pfd[0];
+        m_eventPipeWriteFd = pfd[1];
+        m_eventPipeNotifier = new QSocketNotifier(m_eventPipeReadFd, QSocketNotifier::Read);
+        QObject::connect(m_eventPipeNotifier, &QSocketNotifier::activated, this, &HybrisManager::eventPipeWakeup);
+        m_eventPipeNotifier->setEnabled(true);
+    }
+}
 
-        sensordLogT("HYBRIS EVE %s", sensorTypeName(data.type));
+void HybrisManager::cleanupEventPipe()
+{
+    sensordLogD("cleanup event pipe");
+    if (m_eventPipeNotifier) {
+        delete m_eventPipeNotifier;
+        m_eventPipeNotifier = nullptr;
+    }
+    if (m_eventPipeWriteFd != -1) {
+        ::close(m_eventPipeWriteFd);
+        m_eventPipeWriteFd = -1;
+    }
+    if (m_eventPipeReadFd != -1) {
+        ::close(m_eventPipeReadFd);
+        m_eventPipeReadFd = -1;
+    }
+}
 
-        /* Got data -> Clear the no longer needed fallback event */
-        sensors_event_t *fallback = eventForHandle(data.sensor);
-        if (fallback && fallback->type == data.type && fallback->sensor == data.sensor) {
-            fallback->type = fallback->sensor = 0;
+void HybrisManager::eventPipeWakeup(int fd)
+{
+    if (m_eventPipeReadFd != fd) {
+        m_eventPipeNotifier->setEnabled(false);
+        sensordLogW("fd mismatch, event pipe notifier disabled");
+    } else {
+        const size_t maxEvents = 64;
```

Comment:
These two magic values could be somewhere?

Reply:
Yes, ideally it should be one and same constant.
I started doing that at some stage but gave up as it yielded a lot of diff noise.
Maybe I should do it as a separate commit...

Reply:
Just a define at the top of the file should be simple enough.

Reply:
Yeah, but there are other itches to scratch once one starts...